modules/payment/src/PluginForm/OffsitePaymentForm.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Drupal\commerce_payment\PluginForm;
+
+use Drupal\commerce_payment\Plugin\Commerce\PaymentGateway\OffsiteRedirectPaymentGatewayInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Render\Element;
+
+abstract class OffsitePaymentForm extends PaymentGatewayFormBase {
```

Comment:
I do not think we need a form for each case.

1. Direct redirect: you must do some kind of handshake with the gateway. You send order details, they send a one time checkout link. This has to be performed in the form and `NeedsRedirectException` thrown. No need for its own form class.
2. POST redirect: you'll need to add form items, this covers it.
3. Iframe: covered. If the gateway plugin does not implement `OffsiteRedirectPaymentGatewayInterface` it will include their form, which should have hte iframe code. It does not attempt to redirect.

Each implementation is so wildly different I don't know if we could provide enough boilerplate code to support it. We have _just enough_ now to make working a lot easier.